### DISCUSSION TITLE
How can I slow the frame generation from x to y for like an intro part

### DISCUSSION THREAD

**Discussion by gateway** [rkhamilton/vqgan-clip-generator#16] [Ideas]
Sorry this is hard to describe so a visual aid might help.  Here is something I did a long time ago but never made a video of it.

![bmlogo](https://user-images.githubusercontent.com/519563/137973403-407dcbcd-ca4a-480b-9a3d-3c3cf94e3e5f.jpg)

What Im trying to accomplish is that between iteration 0 to lets say 100 or 150 I want to get like maybe 5 seconds (a slow transformation) before it starting to take shape on its own.  I see a lot of cool stuff but when an image is used as a reference it goes by so fast I really want to see a slow transformation.

Also It might be neat after you hit your x iterations to reverse it back to normal.  This could be manually done.

Or when you hit x iterations (frames) start to zoom in or out, then when zooming in or out slowly go back to iteration 0?

Just thinking of other ways to do more than just a basic animation, zoom, pan, tilt etc.. thoughts?

**Comment by rkhamilton** [comment]
I don't follow exactly what you'd like to do, but I'll discuss a few scenarios:
The easiest / lowest coding approach is to generate multiple image sequences and combine them together in the end (or splice videos in your favorite NLE). For example:
1. Generate sequence 1 with save_every set to a lower number, say 2. This will give you more images for the same amount of training.
2. Take the last image of sequence 1, and use that as init_image for sequence 2. Set a higher save_every, so images will change faster between frames.

Likewise, instead of changing save_every, you could change the zoom rate from 1.0 to 1.02 (to start zooming). 

You may see a discontinuity in the image look because the obvious path of resetting the GAN will lose the training. But this is a straightforward way to set up changes.

You're going to have the most control and creativity by writing your own version of a generate.zoom_video_frames function. I have an example of a [custom zoom video script](https://github.com/rkhamilton/vqgan-clip-generator/blob/main/examples/custom_zoom_video.py) showing how you can easily pull out any of those methods and really get into the guts of how it's done. I wanted to be able to more easily create custom generate methods, which is why I made this package in the first place. If you look at that script you will see that it's still pretty high level, with the guts still buried.

You could code your video transitions in that script by changing the criteria for saving files, changing the saving interval over time, or change the criteria for zooming / shifting to also include a wait for interations to be greater than a threshold.

**Reply by gateway** [reply]
Yea sorry I didn't do a good job of explaining what I was thinking about.. 

The image above from iteration 0 to 150 I would like for it to happen over 4 seconds so I need a lot more in-between fames to fill that part in. So that I understand I would need to set the following.

iterations = 1, // or 2
save_every=x amount, // to get the number of frames I want 

As far as other video style options, I was just saying zooming was kinda played out so I was looking for a more unique control over some of the ideas I have.

1.  Iterate from an image slowly over x seconds
2. Change the iteration rate to a higher value
3. Begin a slow zoom in or out
4. Maybe be able to blend out to another image over iterations.. so for example in the images above during a zoom in that it slowly goes back to iteration 0 as in the photo above.

Its more of a combo of those .. I worry that trying to do a bunch of frames, restart it do another bunch of frames that the flow or the consistency would be changed etc.

Ill look into doing some of your function but wasn't sure If I could do something like this..

```generate.restyle_video_frames(original_video_frames,
        eng_config=config,
        text_prompts = text_prompts,
        iterations = 1,
        save_every=2,
        current_source_frame_prompt_weight=0.1,
        previous_generated_frame_prompt_weight=0.0,
        generated_frame_init_blend=0.1)
 // stop at video frame 800 (just an example)

 // start at video frame 801 with proper seed and learning rate
generate.restyle_video_frames(original_video_frames,
        eng_config=config,
        text_prompts = text_prompts,
        iterations = 25,
        save_every=10,
        current_source_frame_prompt_weight=0.1,
        previous_generated_frame_prompt_weight=0.0,
        generated_frame_init_blend=0.1)
// insert zoom code to start at frame 1000 

etc ...```

**Reply by rkhamilton** [reply]
Your pseudocode reads like my first "easiest / lowest coding approach" example where you are calling the generate function repeatedly. I might expect discontinuities in the look of frames because each call resets the GAN. It might not be noticeable. Note that you'd need to insert a line to set init_image = last_generated_image.png from the previous block, but yeah, that should work. The file numbering will reset, but you can renumber them using any number of tools. Or just convert them into separate clips and combine them in your favorite video editor.

If you instead put together multiple for loops within the same script you could make it do what you describe in your list of ideas. Here's a **non-functional pseudocode** example of how you might think about editing your own generate script. Now the repeated BLOCKs. Each of those BLOCKs creates a series of images with certain settings. Note that in each block of new code you would not reset video_frame_num, so that your file numbering won't reset. Don't call eng.initialize_VQGAN_CLIP() in between blocks, because that would reset the image generation. You would simply duplicate the for iteration_num block and just customize the transitions that happen.
```python
    eng = Engine(eng_config)
    eng.initialize_VQGAN_CLIP()
    eng.encode_and_append_prompts(0, parsed_text_prompts, parsed_image_prompts, parsed_noise_prompts)
    eng.configure_optimizer()

    # generate images
    ## BLOCK 1
    video_frame_num = 1
    for iteration_num in tqdm(range(1,iterations+1),unit='iteration',desc='zoom video'):
        # do all the stuff
        eng.train(iteration_num)
        if save_every and iteration_num % save_every == 0:
            eng.convert_image_to_init_image(last_frame_as_pil_image)

            # Re-create optimiser with the new initial image
            eng.configure_optimizer()

            eng.save_current_output(filepath_to_save)
            video_frame_num += 1
    ## BLOCK 2
    for iteration_num in tqdm(range(1,iterations+1),unit='iteration',desc='zoom video'):
        # do all the stuff but different
        eng.train(iteration_num)
        if save_every and iteration_num % save_every == 0:
            eng.convert_image_to_init_image(last_frame_as_pil_image)

            # Re-create optimiser with the new initial image
            eng.configure_optimizer()

            eng.save_current_output(filepath_to_save)
            video_frame_num += 1
    ## BLOCK 3
    for iteration_num in tqdm(range(1,iterations+1),unit='iteration',desc='zoom video'):
        # do all the stuff even more different
        eng.train(iteration_num)
        if save_every and iteration_num % save_every == 0:
            eng.convert_image_to_init_image(last_frame_as_pil_image)

            # Re-create optimiser with the new initial image
            eng.configure_optimizer()

            eng.save_current_output(filepath_to_save)
            video_frame_num += 1
```

**Reply by gateway** [reply]
nice i may give it a shot this weekend to see what comes up.. btw does seed value you need to be set anywhere to keep it consistent?

And thank you.

**Reply by rkhamilton** [reply]
If you want two runs to be same-ish, you can set it in the VQGAN_CLIP_Config instance.
```python
eng_config=VQGAN_CLIP_Config()
eng_config.seed = 1234
eng = Engine(eng_config)
```
This is basically what I do in my restyle_video loop. Each frame I create a new Engine with the same config (and same seed) as before.

If you want to switch back to the same seed later without reloading the model and resetting the training, you can use Engine.set_seed.
```python
eng.set_seed(1234)
```

Another suggesion: instead of stacking up for iteration loops, if you are only changing parameters you could insert a function to set them based on video_fram_num. E.g.
```python
def parameters_by_frame(frame_num):
        shift_x = 0 if frame_num < 100 else 1
        shift_y = 0 if frame_num < 100 else 1
        zoom_scale = 1.0 if frame_num < 100 else 1.02
        return shift_x, shift_y, zoom_scale


for iteration_num in range(1,iterations+1):
    eng.train(iteration_num)
    if save_every and iteration_num % save_every == 0:
        shift_x, shift_y, zoom_scale = parameters_by_frame(video_frame_num)

        eng.convert_image_to_init_image(last_frame_as_pil_image)

        # Re-create optimiser with the new initial image
        eng.configure_optimizer()

        eng.save_current_output(filepath_to_save)
        video_frame_num += 1
```

Anything that is a simple parameter could easily be done that way by writing a function like that. Of course you could make it more complex, with changes after many different frame numbers, or changing continuously, like shifting in a circle over time, etc.
